### PR TITLE
Add architecture documentation about runtime & device compiler

### PR DESCRIPTION
**Pull request by keryell**
Add also some hints about using the device compiler with PoCL and Xilinx FPGA.

### CODE REVIEW

**Comment by MathiasMagnus** [comment]
Wow, even I, who has some clue about the internals have no idea what the triSYCL workflow diagram is supposed to mean.

File: doc/architecture.rst [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+The compilation flow to generate the final ``.kernel_caller.ll`` file
+is based on LLVM ``opt`` to apply a sequence of LLVM passes:
+
+``-globalopt -deadargelim``
+  to clean-up the code before SYCL massaging;
+
+``-SYCL-args-flattening``
+  is a fundamental SYCL-specific pass that takes the lambda capture
+  (basically a C++ structure passed by address) of a SYCL kernel
+  lambda expression and flatten it as its content. So basically if the
```
Comment (yu810226) [inline doc/architecture.rst]: Should it be flattens?

File: doc/architecture.rst [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+  capture has several scalar and accessor parameters, the structure
+  address used in the function call is replaced by a function call
+  with all the parameters explicitly passed as argument. This makes
```
Comment (yu810226) [inline doc/architecture.rst]: arguments?

File: doc/architecture.rst [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+     cl::sycl::drt::serialize_arg(detail::task &task, std::size_t index,
+                                  void *arg, std::size_t arg_size);
+     // or for an accessor argument
+     cl::sycl::drt::serialize_accessor_arg(detail::task &task, std::size_t index,
+                                           void *arg, std::size_t arg_size)
+
+  The marking functions generated by triSYCL headers are in
+  `<include/CL/sycl/detail/instantiate_kernel.hpp>`_ while the
+  functions used by the transformed code are in
+  `<include/CL/sycl/device_runtime.hpp>`_. The functions from
```
Comment (yu810226) [inline doc/architecture.rst]: These 2 links are not working.

File: doc/architecture.rst [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+
+     cl::sycl::drt::set_kernel(detail::task &task, const char *kernel_name,
+                               const char *kernel_short_name);
+     // For each parameter call:
+     // either for a scalar argument
+     cl::sycl::drt::serialize_arg(detail::task &task, std::size_t index,
+                                  void *arg, std::size_t arg_size);
+     // or for an accessor argument
+     cl::sycl::drt::serialize_accessor_arg(detail::task &task, std::size_t index,
+                                           void *arg, std::size_t arg_size)
```
Comment (yu810226) [inline doc/architecture.rst]: There is a missing `;`.

File: doc/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+
+  The marking functions generated by triSYCL headers are in
+  `<include/CL/sycl/detail/instantiate_kernel.hpp>`_ while the
+  functions used by the transformed code are in
+  `<include/CL/sycl/device_runtime.hpp>`_. The functions from
+  ``cl::sycl::drt::`` are the link to the underlying runtime, such as
```
Comment (yu810226) [inline doc/architecture.rst]: links?

File: doc/architecture.rst [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The file extensions used on the kernel side are:
+
+``.pre_kernel.ll``
+  the SYCL C++ code compiled by Clang for the host side, including
+  the call of the kernels;
+
+``.kernel.ll``
+  the LLVM IR of the host code after the LLVM triSYCL pass
```
Comment (yu810226) [inline doc/architecture.rst]: the kernel code.

Comment (yu810226) [reply doc/architecture.rst]: Now we only get kernel.bc. Not .ll. But it is the same meaning anyway.

**Comment by keryell** [comment]
@MathiasMagnus because it is about the device branch we do not have merged yet, so probably you have never got lost into this area. :-)

But we are thinking to merge soon. Even if it is even more experimental than the CPU support, it is interesting from a research perspective. And it will be easier with 1 main branch only. :-)

File: doc/architecture.rst [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+Usage
+-----
+
+Unfortunately there is no driver yet to generate directly the host and
+device part and it is up to the end-user for now, since it is still
+experimental and in development. So using the compiler
+is... painful. :-(
+
+It is expected to be used as for example with examples from
+`<../tests/device_compiler>`_. Everything is done from
```
[the remark this answers is not in the document]
Comment (keryell) [reply doc/architecture.rst]: Yes, I have to merge the device branch first

File: doc/architecture.rst [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+
+Using Xilinx SDx xocc for FPGA
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's assume you have installed Xilinx SDx somewhere. Initialize the
+environment with something like::
+
+  export XILINX_SDX=/opt/Xilinx/SDx/2017.2
+  PATH=$PATH:$XILINX_SDX/bin
+
```
[the remark this answers is not in the document]
Comment (keryell) [reply doc/architecture.rst]: Good point!

File: doc/architecture.rst (hunk 1 above)
Comment (keryell) [reply doc/architecture.rst]: Yes.

File: doc/architecture.rst (hunk 2 above)
Comment (keryell) [reply doc/architecture.rst]: Right

File: doc/architecture.rst (hunk 6 above)
Comment (keryell) [reply doc/architecture.rst]: Yes but you are right, it is better if it fits the real life :-)

**Comment by yu810226** [comment]
Note for this PR: the link issue can be solved after merging the device branch.

**Comment by yu810226** [comment]
Thanks! :-)))))

**Pull request merged**